### ISSUE TITLE
Incorrect display of the command shell prompt after exiting the Vim program

### ISSUE THREAD

**Issue by RestorerZ** [vim/vim#12430]
### Steps to reproduce

1. Open the standard Windows Command shell --- cmd.exe
2. For illustrative purposes, fill the screen with some text, for example, enter the command `help` in the command shell prompt.
3. Run the `Vim ‐u NONE --noplugin`
4. After that, immediately exit the Vim program with the command `:q`
5. The command shell prompt will be displayed at the top of the command shell window over the text already there.

![not_correct_promt-cmd12c02b9ec6d29c5d6.png](https://imgbb.su/images/2023/05/23/not_correct_promt-cmd12c02b9ec6d29c5d6.png)

This is clearly visible in the attached screenshot.
The starting position of the invitation before launching Vim is marked with a blue stripe.
The position of the prompt after exiting Vim is marked with a red stripe.

This behavior of Vim has been observed since version [9.0.1065](https://github.com/vim/vim-win32-installer/releases/tag/v9.0.1065).
Versions [9.0.1060](https://github.com/vim/vim-win32-installer/releases/tag/v9.0.1060) and earlier worked correctly.

I think something is broken either in the patch [9.0.1063](https://github.com/vim/vim/releases/tag/v9.0.1063) or in the patch [9.0.1065](https://github.com/vim/vim/releases/tag/v9.0.1065).

### Expected behaviour

The command prompt is displayed at the bottom of the screen and does not overlap the existing text.
As in this screenshot.

![correct_promt-cmd1fd287fa150c79e25.png](https://imgbb.su/images/2023/05/23/correct_promt-cmd1fd287fa150c79e25.png)

### Version of Vim

9.0.1572

### Environment

WindowsNT 6.2

**Comment by brammool** [comment]
> ### Steps to reproduce
> 
> 1. Open the standard Windows Command shell --- cmd.exe
> 2. For illustrative purposes, fill the screen with some text, for example, enter the command `help` in the command shell prompt.
> 3. Run the `Vim ‐u NONE --noplugin`
> 4. After that, immediately exit the Vim program with the command `:q`
> 5. The command shell prompt will be displayed at the top of the command shell window over the text already there.
> 
> ![not_correct_promt-cmd12c02b9ec6d29c5d6.png](https://imgbb.su/images/2023/05/23/not_correct_promt-cmd12c02b9ec6d29c5d6.png)
> 
> This is clearly visible in the attached screenshot.
> The starting position of the invitation before launching Vim is marked with a blue stripe.
> The position of the prompt after exiting Vim is marked with a red stripe.
> 
> This behavior of Vim has been observed since version [9.0.1065](https://github.com/vim/vim-win32-installer/releases/tag/v9.0.1065).
> Versions [9.0.1060](https://github.com/vim/vim-win32-installer/releases/tag/v9.0.1060) and earlier worked correctly.
> 
> I think something is broken either in the patch [9.0.1063](https://github.com/vim/vim/releases/tag/v9.0.1063) or in the patch [9.0.1065](https://github.com/vim/vim/releases/tag/v9.0.1065).

What is the value of the 'term' option?

Other options that matter:
	set t_te? t_TE?

Have you tried using a console window?

You could try to undo patch 9.0.1063, which is mainly moving this line
further down:

	out_str(T_TE);

However, that cannot be the solution, since it will break other
terminals (it was changed for use with Kitty).  Perhaps using "#ifdef
MSWIN" can be used, but there are likely problems when using something
else than the standard Windows console or terminal.


-- 
hundred-and-one symptoms of being an internet addict:
63. You start using smileys in your snail mail.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by RestorerZ** [comment]
> What is the value of the 'term' option?

`term=win32`


> Other options that matter: set t_te? t_TE?

`t_te=^[|E`


> Have you tried using a console window?

Yes.

I also tried running Vim in the following console emulators:
[ConEmu](https://github.com/Maximus5/ConEmu) and [cmder](https://github.com/cmderdev/cmder) based on it --- work is normal there. The command prompt is displayed correctly.
When the Vim is running in a PowerShell window that uses cmd and conhost as console emulators, the command prompt is displayed incorrectly after exiting Vim.
The values of the parameters `term` and `t_te` are the same everywhere, such as those shown above.

> 
```
:set termcap
--- Параметры терминала ---
  t_AB=^[|%db         t_ds=               t_RC=               t_TI=
  t_AF=^[|%df         t_Ds=               t_RI=               t_Ts=
  t_AU=               t_EC=               t_Ri=               t_ts=
  t_AL=^[|%dL         t_EI=               t_RK=               t_u7=
  t_al=^[|L           t_fs=               t_RS=               t_ue=^[|0m
  t_bc=               t_fd=               t_RT=               t_us=^[|67m
  t_BE=               t_fe=               t_RV=               t_Us=
  t_BD=               t_GP=               t_Sb=               t_ut=y
  t_cd=^[|j           t_IE=               t_SC=               t_vb=^[|B
  t_ce=^[|K           t_IS=               t_se=^[|0m          t_ve=^[|V
  t_Ce=               t_ke=               t_Sf=               t_vi=^[|v
  t_cl=^[|J           t_ks=               t_SH=               t_VS=
  t_cm=^[|%i%d;%dH    t_le=^H             t_SI=               t_vs=
  t_Co=16             t_mb=               t_Si=               t_WP=
  t_CS=               t_md=^[|15m         t_so=^[|31m         t_WS=
  t_Cs=               t_me=^[|7m          t_SR=               t_XM=
  t_cs=^[|%i%d;%dr    t_mr=^[|112m        t_sr=               t_xn=y
  t_CV=^[|%d;%dV      t_ms=y              t_ST=               t_xs=
  t_da=               t_nd=               t_Te=               t_ZH=^[|225m
  t_db=               t_op=               t_te=^[|E           t_ZR=^[|0m
  t_DL=^[|%dM         t_RF=               t_TE=
  t_dl=^[|M           t_RB=               t_ti=^[|S
  t_8f=^[[38;2;%lu;%lu;%lum
  t_8b=^[[48;2;%lu;%lu;%lum
  t_8u=^[[58;2;%lu;%lu;%lum
--- Клавиши, доступные в терминале ---
t_@7 <End>       <ce>^CO   t_kB <S-Tab>     <ce>^C^O       <S-F3>      <ce>^CV
t_K6 <kPlus>     <ce>^CN   t_kD <Del>       <ce>^CS        <S-F4>      <ce>^CW
t_K7 <kMinus>    <ce>^CJ   t_kI <Insert>    <ce>^CR        <S-F5>      <ce>^CX
t_K9 <kMultiply> <ce>^C7   t_kN <PageDown>  <ce>^CQ        <S-F6>      <ce>^CY
t_k1 <F1>        <ce>^C;   t_kP <PageUp>    <ce>^CI        <S-F7>      <ce>^CZ
t_k2 <F2>        <ce>^C<   t_kb <BS>        <ce>^Cx        <S-F8>      <ce>^C[
t_k3 <F3>        <ce>^C=   t_kd <Down>      <ce>^CP        <S-F9>      <ce>^C\
t_k4 <F4>        <ce>^C>   t_kh <Home>      <ce>^CG        <S-F10>     <ce>^C]
t_k5 <F5>        <ce>^C?   t_kl <Left>      <ce>^CK        <C-Left>    <ce>^Cs
t_k6 <F6>        <ce>^C@   t_kr <Right>     <ce>^CM        <C-Right>   <ce>^Ct
t_k7 <F7>        <ce>^CA   t_ku <Up>        <ce>^CH        <C-Home>    <ce>^Cw
t_k8 <F8>        <ce>^CB        <Mouse>     <9b>M          <C-End>     <ce>^Cu
t_k9 <F9>        <ce>^CC        <S-F1>      <ce>^CT                    <ce>^Cy
t_k; <F10>       <ce>^CD        <S-F2>      <ce>^CU
t_#2 <S-Home>    <ce>^C<c2>             t_KG <k4>        <ce>^C<ea>
t_#4 <S-Left>    <ce>^C<c9>             t_KH <k5>        <ce>^C<ee>
t_%i <S-Right>   <ce>^C<cb>             t_KI <k6>        <ce>^C<f2>
t_*7 <S-End>     <ce>^C<cd>             t_KJ <k7>        <ce>^C<f6>
t_F1 <F11>       <ce>^C<85>             t_KK <k8>        <ce>^C<fa>
t_F2 <F12>       <ce>^C<86>             t_KL <k9>        <ce>^C<fe>
t_KC <k0>        <ce>^C<da>                  <S-Up>      <ce>^C<c4>
t_KD <k1>        <ce>^C<de>                  <S-Down>    <ce>^C<cf>
t_KE <k2>        <ce>^C<e2>                  <S-F11>     <ce>^C<87>
t_KF <k3>        <ce>^C<e6>                  <S-F12>     <ce>^C<88>
Чтобы продолжить, нажмите клавишу <ENTER> или введите команду
```
This is the output of the `set termcap` command in the Windows cmd window

**Comment by RestorerZ** [comment]
Bram, one more detail, maybe it will be useful.
If you switch to edit mode in a newly launched Vim in a new empty buffer, type text, exit edit mode and enter the command `:q` and to the request "Save changes to "No file"?" If you answer "No", then after exiting Vim, the command shell prompt is displayed correctly.
In all other cases:
1. Open Vim and immediately exit it by command `:q`;
2. Open an existing file in Vim, edit it, save it and exit Vim;
3. Save the changes to a new buffer and exit Vim.
In all these cases, the command prompt will be displayed incorrectly.

**Comment by brammool** [comment]
The "Save changes" dialog will move the cursor to the command line.  However, switching to the alternate screen would undo this, I thought it has its own cursor position.  Not sure why the dialog makes it work OK.
Moving the cursor after switching screens is problematic, since we don't know where to move it to.  I'm not sure it works to move it to the last line of the window (possibly leaving a gap with unused lines, but that is better than putting the prompt halfway previous output).

**Comment by peterroshouse** [comment]
This thread title is exactly right for what I'm seeing though not on Windows (Suse Linux). Perhaps it's related. 

Run vim, edit a file or not and simply exit with :q and the first 2 characters of my prompt are missing. Press enter once and the prompt is normal again.

I found that if I set my prompt PS1 variable to something that has embedded color codes such as below, the prompt is not impacted. For now this is my workaround.
PS1="\[\e[1;32m\]\h: \$ \[\e[0m\] "

Current version of vim is 9.0.1386-17.15.4. 
Downgraded to 9.0.1234-17.12.1 and saw the same issue. 
Downgraded to 9.0.0814-17.9.1 and the issue went away.
